### ISSUE TITLE
Adding modifiers to a logical roll loses the metadata

### ISSUE THREAD

**Issue by saif-ellafi** [Adventuresmith/dart-dice-parser#1]
Hi, thank you for this awesome library (I sent you an email by the way), many of our users will begin using this indirectly through our app :)

For example `/r 2d20kh` gives metadata information, but `/r 2d20kh + 1` does not. It would be great to drag along the metadata perhaps, when a modifier (+ 1) is also used?

Cheers

**Comment by stevesea** [comment]
From my vague recollection of what I was thinking... When the graph is traversed to process the dice expression, there seemed to be two distinct types of results -- either a bag o' dice-roll results + metadata, or a simple number. 

there were a few places where I was _only_ interested in the numeric result and not the dice results
it felt like allowing certain dice operations after an arithmetic operation ("+1", or "#") didn't make much sense. 

e.g.  how does the result after "+1" effect a further evaluation in expressions like `(3d6 + 1 ) kh` or `(3d6 +1) #` ? 

Feels like it those example expressions should be errors. But, in hindsight, letting that complication affect the metadata that's carried up the graph was probably a bad choice.

**Comment by stevesea** [comment]
I misdiagnosed things in my earlier comment (it's not really related to the way that counting collapses a set of results into a single result).

Here's the direction I'm heading... I made some major changes to how the results are structured.

This example output should give you an idea of where I'm heading. 

```console
❯ dart run example/main.dart -r 1234 -o pretty '2d20 kh + 1'
(((2d20) kh ) + 1) ===> RollSummary(total: 17, results: [16, 1], metadata: {rolled: [4, 16], discarded: [4]})
  (((2d20) kh ) + 1) =add=> RollResult(total: 17, results: [16, 1])
      ((2d20) kh ) =drop=> RollResult(total: 16, results: [16], metadata: {discarded: [4]})
          (2d20) =rollDice=> RollResult(total: 20, results: [4, 16], metadata: {rolled: [4, 16]})


❯ dart run example/main.dart -r 1234 -o pretty '2d20 kh '
((2d20) kh ) ===> RollSummary(total: 16, results: [16], metadata: {rolled: [4, 16], discarded: [4]})
  ((2d20) kh ) =drop=> RollResult(total: 16, results: [16], metadata: {discarded: [4]})
      (2d20) =rollDice=> RollResult(total: 20, results: [4, 16], metadata: {rolled: [4, 16]})

```

Basically... the return value of roll() will be a RollSummary. It'll consolidate some of the data from the result nodes. But, they'll also be returned so a client can do... other things? Not sure what might be useful to you or how you're using the data.


Similarly, the RollSummary will include a rollup of the success/failures:
```console
❯ dart run example/main.dart -r 1234 -o pretty '(4d6 + 4d6!) kh3 #cf #cs'
(((((4d6) + ((4d6) ! )) kh 3) #cf ) #cs ) ===> RollSummary(total: 16, results: [6, 5, 5], metadata: {rolled: [6, 2, 1, 5, 3, 5, 1, 4], discarded: [4, 3, 2, 1, 1], score: {critSuccesses: [6]}})
  (((((4d6) + ((4d6) ! )) kh 3) #cf ) #cs ) =count=> RollResult(total: 16, results: [6, 5, 5], metadata: {score: {critSuccesses: [6]}})
      ((((4d6) + ((4d6) ! )) kh 3) #cf ) =count=> RollResult(total: 16, results: [6, 5, 5])
          (((4d6) + ((4d6) ! )) kh 3) =drop=> RollResult(total: 16, results: [6, 5, 5], metadata: {discarded: [4, 3, 2, 1, 1]})
              ((4d6) + ((4d6) ! )) =add=> RollResult(total: 27, results: [1, 1, 2, 3, 4, 5, 5, 6])
                  (4d6) =rollDice=> RollResult(total: 14, results: [6, 2, 1, 5], metadata: {rolled: [6, 2, 1, 5]})
                  ((4d6) ! ) =explode=> RollResult(total: 13, results: [3, 5, 1, 4])
                      (4d6) =rollDice=> RollResult(total: 13, results: [3, 5, 1, 4], metadata: {rolled: [3, 5, 1, 4]})

# in the above example, notice the interaction between the placement of the keep-high vs. critical failure.
# keep high drops the 1's before the results are scored, so no critical failures are recorded in the above.

❯ dart run example/main.dart -r 1234 -o pretty '(4d6 + 4d6!)  #cf #cs'
((((4d6) + ((4d6) ! )) #cf ) #cs ) ===> RollSummary(total: 27, results: [6, 2, 1, 5, 3, 5, 1, 4], metadata: {rolled: [6, 2, 1, 5, 3, 5, 1, 4], score: {critSuccesses: [6], critFailures: [1, 1]}})
  ((((4d6) + ((4d6) ! )) #cf ) #cs ) =count=> RollResult(total: 27, results: [6, 2, 1, 5, 3, 5, 1, 4], metadata: {score: {critSuccesses: [6]}})
      (((4d6) + ((4d6) ! )) #cf ) =count=> RollResult(total: 27, results: [6, 2, 1, 5, 3, 5, 1, 4], metadata: {score: {critFailures: [1, 1]}})
          ((4d6) + ((4d6) ! )) =add=> RollResult(total: 27, results: [6, 2, 1, 5, 3, 5, 1, 4])
              (4d6) =rollDice=> RollResult(total: 14, results: [6, 2, 1, 5], metadata: {rolled: [6, 2, 1, 5]})
              ((4d6) ! ) =explode=> RollResult(total: 13, results: [3, 5, 1, 4])
                  (4d6) =rollDice=> RollResult(total: 13, results: [3, 5, 1, 4], metadata: {rolled: [3, 5, 1, 4]})

❯ dart run example/main.dart -r 1234 -o pretty '(4d6 + 4d6!)  #cf #cs #s>=5 #f<=2'
((((((4d6) + ((4d6) ! )) #cf ) #cs ) #s>= 5) #f<= 2) ===> RollSummary(total: 27, results: [6, 2, 1, 5, 3, 5, 1, 4], metadata: {rolled: [6, 2, 1, 5, 3, 5, 1, 4], score: {successes: [6, 5, 5], failures: [2, 1, 1], critSuccesses: [6], critFailures: [1, 1]}})
  ((((((4d6) + ((4d6) ! )) #cf ) #cs ) #s>= 5) #f<= 2) =count=> RollResult(total: 27, results: [6, 2, 1, 5, 3, 5, 1, 4], metadata: {score: {failures: [2, 1, 1]}})
      (((((4d6) + ((4d6) ! )) #cf ) #cs ) #s>= 5) =count=> RollResult(total: 27, results: [6, 2, 1, 5, 3, 5, 1, 4], metadata: {score: {successes: [6, 5, 5]}})
          ((((4d6) + ((4d6) ! )) #cf ) #cs ) =count=> RollResult(total: 27, results: [6, 2, 1, 5, 3, 5, 1, 4], metadata: {score: {critSuccesses: [6]}})
              (((4d6) + ((4d6) ! )) #cf ) =count=> RollResult(total: 27, results: [6, 2, 1, 5, 3, 5, 1, 4], metadata: {score: {critFailures: [1, 1]}})
                  ((4d6) + ((4d6) ! )) =add=> RollResult(total: 27, results: [6, 2, 1, 5, 3, 5, 1, 4])
                      (4d6) =rollDice=> RollResult(total: 14, results: [6, 2, 1, 5], metadata: {rolled: [6, 2, 1, 5]})
                      ((4d6) ! ) =explode=> RollResult(total: 13, results: [3, 5, 1, 4])
                          (4d6) =rollDice=> RollResult(total: 13, results: [3, 5, 1, 4], metadata: {rolled: [3, 5, 1, 4]})

```

it felt like an app would find it useful to know _which_ rolls were success/failures/crits so it could render them differently if it fit the app's use case.

**Comment by stevesea** [comment]
or, maybe this diagram will illustrate the result graph better than the blocks of text above

```mermaid
flowchart TD
   
    DROP --> ADD;
    ADD --> ROLLB;
    ADD --> EXPLODE;
    EXPLODE --> ROLLA;

    SUMMARY ---|details| DROP_RESULT;
    DROP_RESULT -->|left| ADD_RESULT;
    ADD_RESULT -->|left| ROLLB_RESULT;
    ADD_RESULT -->|right| EXPLODE_RESULT;
    EXPLODE_RESULT -->|left| ROLLA_RESULT;
    
    ROLLA -.->|#1| ROLLA_RESULT@{ shape: doc, label: "RollResult <br/><br/>results: 2,3,6<br/>rolled:2,3,6" };
    EXPLODE -.->|#2| EXPLODE_RESULT@{ shape: doc, label: "RollResult <br/><br/>results: 2,3,6,4<br/>rolled:4"};
    ROLLB -.->|#3| ROLLB_RESULT@{ shape: doc, label: "RollResult <br/><br/>results: 1,5,2<br/>rolled:1,5,2"};
    ADD -.->|#4| ADD_RESULT@{ shape: doc, label: "RollResult <br/><br/>results: 2,3,6,4,1,5,2"};
    DROP -.->|#5| DROP_RESULT@{ shape: doc, label: "RollResult<br/><br/>results: 6,5,4<br/>discarded: 2,3,1,2"};
    
    DROP(["(3d6 + 3d6!) kh 3"]);
    ADD(["(3d6 + 3d6!)"]);
    ROLLB(["3d6"]);
    EXPLODE(["(3d6)!"]);
    ROLLA(["3d6"]);
    SUMMARY@{ shape: doc, label: "RollSummary<br/><br/>Total: 15 <br/>results:6,5,4<br/><br/>rolled: 2,3,6,4,1,5,2 <br/>discarded: 2,3,1,2" };
```

**Comment by ELS2MU** [comment]
@stevesea woah my goodness, I love your workstyle. Can't be happier to have the opportunity to be around this issue ticket :)

**Comment by saif-ellafi** [comment]
@stevesea woah my goodness, I love your workstyle. Can't be happier to have the opportunity to be around this issue ticket :)
I gladly shared with our app users the awesomeness that comes directly through here. Will send you over a free copy of our app!

**Comment by stevesea** [comment]
fixed in v7.0.0
